sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/http/PathTemplateParser.scala

```diff
@@ -0,0 +1,196 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kalix.springsdk.impl.http
+
+import java.util.regex.Pattern
+import scala.annotation.nowarn
+import scala.util.matching.Regex
+import scala.util.parsing.combinator.Parsers
+import scala.util.parsing.input.{ CharSequenceReader, Positional }
+
+object PathTemplateParser extends Parsers {
+
+  override type Elem = Char
+
+  final class ParsedTemplate(val path: String, template: Template) {
+    val regex: Regex = {
+      def doToRegex(builder: StringBuilder, segments: List[Segment], matchSlash: Boolean): StringBuilder =
+        segments match {
+          case Nil => builder // Do nothing
+          case head :: tail =>
+            if (matchSlash) {
+              builder.append('/')
+            }
+
+            head match {
+              case LiteralSegment(literal) =>
+                builder.append(Pattern.quote(literal))
+              case SingleSegmentMatcher =>
+                builder.append("[^/:]*")
+              case MultiSegmentMatcher() =>
+                builder.append(".*")
+              case VariableSegment(_, None) =>
+                builder.append("([^/:]*)")
+              case VariableSegment(_, Some(template)) =>
+                builder.append('(')
+                doToRegex(builder, template, matchSlash = false)
+                builder.append(')')
+            }
+
+            doToRegex(builder, tail, matchSlash = true)
+        }
+
+      val builder = doToRegex(new StringBuilder, template.segments, matchSlash = true)
+
+      template.verb
+        .foldLeft(builder) { (builder, verb) =>
+          builder.append(':').append(Pattern.quote(verb))
+        }
+        .toString()
+        .r
+    }
+
+    val fields: List[TemplateVariable] = {
+      var found = Set.empty[List[String]]
+      template.segments.collect {
+        case v @ VariableSegment(fieldPath, _) if found(fieldPath) =>
+          throw PathTemplateParseException("Duplicate path in template", path, v.pos.column + 1)
+        case VariableSegment(fieldPath, segments) =>
+          found += fieldPath
+          TemplateVariable(
+            fieldPath,
+            segments.exists(_ match {
+              case ((_: MultiSegmentMatcher) :: _) | (_ :: _ :: _) => true
+              case _                                               => false
+            }))
+      }
+    }
+  }
+
+  final case class TemplateVariable(fieldPath: List[String], multi: Boolean)
+
+  final case class PathTemplateParseException(msg: String, path: String, column: Int)
```

Comment:
It seems that this class has been copied before. Have a look at `PathPatternParser` also in the Spring SDK module.

Reply:
Thanks, looking at merging both now.. I think I've seen it before now that you mentioned but never thought of it again. 💃